### PR TITLE
TensorFlow end model bindings refactor + fixes

### PR DESCRIPTION
**Pull request by ajratner**
Main changes:
- Fixes the grid search issue, but more broadly refactors + fixes model saving / loading (now, any keyword arguments passed into `_build` will automatically be saved / loaded for reconstructing the model.  These are referred to as `model_kwargs` and include anything needed to build the model and/or apply it at test time; this is opposed to `training_kwargs` which are only needed to train the model
- Creates a new graph & session for each training session to keep graphs separate and clean, fixing e.g. `GridSearch`
- Moves training loop into parent class so more code reuse, easier to plug in new models
- General fixes + refactoring
- Adds a new test (notebook)
- Reorganizes contrib dir, all models in there (open to other options here)

### CODE REVIEW

**Comment by ajratner** [comment]
Unfortunately the changes were extensive and so, along with the renaming, the diffs are not all appropriately showing up... happy to discuss this PR in person / over slack to make easier to read.  Shouldn't change much in terms of existing functionality though!

**Comment by ajratner** [comment]
One thing I'd love review feedback on: the scoring functions.  In trying to get it to pass tests, I wrote a lighter-weight `score` function which just computes the score--and assumes that the candidates and labels are collated, thus requires no `session` object (per #569 )--and renaming the old `score` function as `error_analysis`.
* One question: does this make sense / seem like right way to go?
* Also, we should probably make sure the generative and discriminative models are in sync on score functions otherwise will be confusing... not sure what right way to do this is

**Comment by ajratner** [comment]
One other thing: the scores in the intro tutorial after grid search seem pretty low.  This might be a new error I introduced (?) Or just a function of overfitting to a minuscule dev set.  Don't know whether we want to fix this in this PR or after putting in the new dataset.

**Comment by ajratner** [comment]
@stephenbach @jason-fries @henryre ready for review :D!

**Comment by ajratner** [comment]
@stephenbach @jason-fries ready for review again!
